### PR TITLE
Fix/duplicate txn in group

### PR DESCRIPTION
**Pull request by vuvoth**
## Proposed Changes

+ Throw error when send duplicate transaction in one group. 
+ Update and and more unit test.

### CODE REVIEW

File: packages/runtime/test/integration/global-opcodebudget.ts [hunk 1]
```diff
@@ -96,6 +96,10 @@ describe("TEALv6: Global Opcode Budget", function () {
 
 			const buffTx = {
 				...txnParam,
+				payFlags: {
```
Comment (thdailong) [inline packages/runtime/test/integration/global-opcodebudget.ts:99]: Could you explain for me about the payFlags?

Comment (vuvoth) [reply packages/runtime/test/integration/global-opcodebudget.ts:99]: Please check this document. https://algobuilder.dev/guide/execute-transaction.html

File: CHANGELOG.md [hunk 2]
```diff
@@ -117,7 +117,7 @@ recommended to use the method from `Deployer` rather than the function dirrectly
 - Allow token to be empty.
 - Throw error when issue inner transactions in clear program. Fixed in [#667](https://github.com/scale-it/algo-builder/pull/667).
 - Parameters in `extract*` opcodes can greater than uint8. Fixed in [#666](https://github.com/scale-it/algo-builder/pull/666).
-
+- Cannot send duplicate transaction in group transaction.
```
Comment (rmeena840) [inline CHANGELOG.md:120]:
```suggestion
- Restirct duplicate transaction in group transaction. 
```

File: packages/runtime/src/runtime.ts [hunk 3]
```diff
@@ -226,6 +226,15 @@ export class Runtime {
 		}
 	}
 
+	validateDupTxn(gtxns: EncTx[]): void {
+		for (const txn of gtxns) {
+			const count = gtxns.filter((anotherTxn) => anotherTxn.txID === txn.txID).length;
```
Comment (rmeena840) [inline packages/runtime/src/runtime.ts:231]: We can give better name here. `count` of what?

File: packages/runtime/test/integration/pooled-op-cost.ts [hunk 4]
```diff
@@ -50,29 +50,41 @@ describe("TEALv5: Pooled Opcode Cost calculation", function () {
...
 		expectRuntimeError(
 			() =>
 				runtime.executeTx([
 					appCallParam,
-					{ ...appCallParam, appArgs: [`str:${STR_NORMAL_COST}`] },
-					{ ...appCallParam, appArgs: [`str:${STR_NORMAL_COST}`] },
-					{ ...appCallParam, appArgs: [`str:${STR_NORMAL_COST}`] },
+					{
+						...appCallParam,
+						appArgs: [`str:${STR_NORMAL_COST}`],
```
Comment (rmeena840) [inline packages/runtime/test/integration/pooled-op-cost.ts:60]: This line is same in all objects in this array. Maybe we can include that itself in `appCallParam`

File: packages/runtime/test/integration/pooled-op-cost.ts [hunk 5]
```diff
@@ -50,29 +50,41 @@ describe("TEALv5: Pooled Opcode Cost calculation", function () {
...
+		// enough normal cost transactions in group, add note to make txn different with each other
 		const passTxGroup = [
 			appCallParam,
-			{ ...appCallParam, appArgs: [`str:${STR_NORMAL_COST}`] },
-			{ ...appCallParam, appArgs: [`str:${STR_NORMAL_COST}`] },
-			{ ...appCallParam, appArgs: [`str:${STR_NORMAL_COST}`] },
-			{ ...appCallParam, appArgs: [`str:${STR_NORMAL_COST}`] },
-			{ ...appCallParam, appArgs: [`str:${STR_NORMAL_COST}`] },
-			{ ...appCallParam, appArgs: [`str:${STR_NORMAL_COST}`] },
+			{ ...appCallParam, appArgs: [`str:${STR_NORMAL_COST}`], payFlags: { note: "salt 0" } },
```
Comment (rmeena840) [inline packages/runtime/test/integration/pooled-op-cost.ts:82]: ditto

**Review by rmeena840 — CHANGES_REQUESTED** [review]
Reviewed

File: packages/runtime/test/integration/pooled-op-cost.ts (hunk 4 above)
Comment (vuvoth) [reply packages/runtime/test/integration/pooled-op-cost.ts:60]: No, it's different object. I overwrite payFlags and appCallParam also use in other test, so we should not change it.

File: CHANGELOG.md (hunk 2 above)
Comment (vuvoth) [reply CHANGELOG.md:120]: This case is force. If user send duplicate transaction it's will throw error. I worry use `Restrict` will make user confuse.

Comment (rmeena840) [reply CHANGELOG.md:120]: This is for our own track of changes.

**Review by rmeena840 — COMMENTED** [review]
Reviewed

**Comment by vuvoth** [comment]
> Reviewed

Updated

File: packages/runtime/test/src/runtime.ts [hunk 6]
```diff
@@ -310,6 +310,48 @@ describe("Rounds Test", function () {
...
+			fromAccount: alice.account,
+			toAccountAddr: bob.address,
+			amountMicroAlgos: amount,
+			payFlags: {
+				totalFee: fee,
+			},
+		};
+	});
+
+	it("Should fail send duplicate transaction in group", () => {
```
Comment (sczembor) [inline packages/runtime/test/src/runtime.ts:339]:
```suggestion
	it("Should throw an error when sending duplicate tx in a group", () => {
```

File: packages/runtime/test/src/runtime.ts [hunk 7]
```diff
@@ -310,6 +310,48 @@ describe("Rounds Test", function () {
...
+		const groupTx = [paymentTxn, { ...paymentTxn }];
+
+		expectRuntimeError(
+			() => runtime.executeTx(groupTx),
+			RUNTIME_ERRORS.TRANSACTION.TRANSACTION_ALREADY_IN_LEDGER
+		);
+	});
+
+	it("Should work when add different note filed", () => {
```
Comment (sczembor) [inline packages/runtime/test/src/runtime.ts:348]:
```suggestion
	it("Should not throw an error when add different note filed", () => {
```

File: packages/runtime/src/runtime.ts [hunk 8]
```diff
@@ -226,6 +226,15 @@ export class Runtime {
 		}
 	}
 
+	validateNoDupTxn(gtxns: EncTx[]): void {
+		for (const txn of gtxns) {
+			const isDuplicate = gtxns.filter((anotherTxn) => anotherTxn.txID === txn.txID).length > 1;
```
Comment (sczembor) [inline packages/runtime/src/runtime.ts:231]:
```suggestion
			const isDuplicated = gtxns.filter((anotherTxn) => anotherTxn.txID === txn.txID).length > 1;
```

File: packages/runtime/src/runtime.ts [hunk 9]
```diff
@@ -226,6 +226,15 @@ export class Runtime {
...
+			if (isDuplicate) {
```
Comment (sczembor) [inline packages/runtime/src/runtime.ts:232]:
```suggestion
			if (isDuplicated) {
```

File: packages/runtime/src/runtime.ts [hunk 10]
```diff
@@ -903,7 +912,8 @@ export class Runtime {
 
 		// validate first and last rounds
 		this.validateTxRound(gtxs);
-
+		// validate no duplication txns
```
Comment (sczembor) [inline packages/runtime/src/runtime.ts:915]:
```suggestion
		// checks if the transactions are not duplicated
```

File: packages/runtime/src/runtime.ts [hunk 11]
```diff
@@ -226,6 +226,15 @@ export class Runtime {
 		}
 	}
 
+	validateNoDupTxn(gtxns: EncTx[]): void {
```
Comment (sczembor) [inline packages/runtime/src/runtime.ts:229]: Im thinking if `assertNoDuplicateTransaction` name wouldnt be better. Let me know what do you think

File: packages/runtime/test/src/runtime.ts [hunk 12]
```diff
@@ -310,6 +310,48 @@ describe("Rounds Test", function () {
...
+	const fee = 1000;
+
+	let alice: AccountStoreI;
+	let bob: AccountStoreI;
+
+	let runtime: Runtime;
+	let paymentTxn: types.AlgoTransferParam;
+
+	this.beforeEach(() => {
+		alice = new AccountStore(minBalance * 10n);
```
Comment (sczembor) [inline packages/runtime/test/src/runtime.ts:324]: Maybe we could use `defaultAccounts` here. Its optional

**Review by sczembor — APPROVED** [review]
Pre approving. I left few suggestions which are optional. Thanks for the PR

File: packages/runtime/src/runtime.ts (hunk 11 above)
Comment (vuvoth) [reply packages/runtime/src/runtime.ts]: I updated it, please check PR again!

**Review by sczembor — APPROVED** [review]
LGTM!

**Review by rmeena840 — APPROVED** [review]

**Pull request merged**